isde-project group E

G-P-x opened the pull request.

G-P-x committed:
added buttons 1.Download Result  2.Download Plot

G-P-x committed:
Implement the API for returning a JSON file with the results

G-P-x committed:
link used to get the result changed

G-P-x committed:
Implemented the functionality for returning the plot

asotgiu commented on .vscode/c_cpp_properties.json:
Why is this file versioned?

asotgiu commented on .vscode/launch.json:
Why is this file versioned?

asotgiu commented on .vscode/settings.json:
Why is this file versioned?

asotgiu commented on app/ml/image_transformation.py:
```diff
@@ -0,0 +1,20 @@
+        input_image_path = os.path.join(conf.image_folder_path, image_id)
+        image = Image.open(input_image_path)
+        sharped_image = ImageEnhance.Sharpness(image).enhance(value)
+        sharped_image_id = image_id.replace(image_id, "sharped_image.JPEG")
+        output_image_path = os.path.join(conf.image_folder_path, sharped_image_id)
+        sharped_image.save(output_image_path, format="JPEG")  # create a sharped copy of the original image
```
What happens if multiple users ask for the same job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on app/ml/image_uploader.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def upload_image(contents):
+    image_id = "uploaded_image.JPEG"
+    image_path = Configuration.image_folder_path_img + image_id
+    # Saving the file
+    with open(f"{image_path}", "wb") as f:
+        f.write(contents)
```
What happens if multiple users ask for the same job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on main.py:
```diff
@@ -45,12 +64,30 @@ def create_classify(request: Request):
     model_id = form.model_id
     classification_scores = classify_image(model_id=model_id, img_id=image_id)
+    result.generate_JSON(classification_score= classification_scores)  #generate JSON file for the result of classification-score
+    with open("result.json", "w") as f:
+        json.dump(result.classification_results, f)  #write the result into the JSON file   
```
What happens if multiple users ask for a classification job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on main.py:
```diff
@@ -45,12 +64,30 @@ def create_classify(request: Request):
+
+
+    # Save the chart as a PNG file
+    plt.savefig('plot.png')
```
What happens if multiple users ask for a classification job? The file would be overwritten. Isn't it best to return the results directly? Or, if you reuse them later for other things, another way to do it would be to store the results with a unique name and clean up after the response is returned to the user (or after a while anyway).

asotgiu commented on main.py:
```diff
@@ -59,3 +96,72 @@ async def request_classification(request: Request):
+
+@app.post("/image_from_PC")
+async def create_upload_image(request: Request, file: UploadFile = File(...)): 
+    contents = await file.read()
+    image_id = upload_image(contents)
+    form = ClassificationForm(request)
+    await form.load_data()
+    model_id = form.model_id
+    classification_scores = classify_image(model_id=model_id, img_id=image_id)
```
If something goes wrong (for instance, an invalid file is uploaded), the app crashes and the file remains in memory anyway. The error should be handled (for instance with a try-except) and the file should be removed. Moreover, the user does not know what happened as no specific error message is shown in the web interface.

asotgiu commented on plot.png:
Why is this file versioned?

asotgiu commented on result.json:
Why is this file versioned?

asotgiu reviewed (changes requested):
I reviewed the code and tested the app. Most of the requested functionalities are implemented, but there are some issues to solve before accepting the Pull Request.
Please fix them and request a new review when it's done. There's no need to open a new PR, it's enough to push the new commits to the origin repository (yours) of the PR. In that way, I will see only the new changes and I won't have to review the full code again.
- Issue #2 required to modify image color, brightness, contrast, and sharpness. However, only the latter is implemented. Please, add the other transformations.
- Remove files that should not be versioned, or provide an explanation on why the file is there (perhaps they should just be added to the .gitignore).
- Fix the concurrency problem with the results, the plots, and the uploaded images being stored with the same name, and handle the uploading of invalid images.
- Improve the documentation through the code and add a changelog with the list of resolved issues to the pull request documentation (edit the description of the PR, or at least use the GUI functionality to link the PR to the issues of the target repository).

Thanks!

G-P-x committed:
removed unnecessary lines

G-P-x committed:
added a button and a javascript script to download the graph

G-P-x committed:
added a button and a javascript script in order to download the scores in a json file

G-P-x committed:
removed the python fucntion to return the png and the json

G-P-x committed:
removed all non necessary lines, releated to the result variable, about returning the png and the json from the main

G-P-x committed:
deleted plot.png and result.json from the project, they're not needed

G-P-x committed:
deleted data_storage.py, the class are not longer needed with the new implementation for downloading the png and the json

G-P-x committed:
moved the javascript script for returning the png from the html to a new file in the static folder

G-P-x committed:
moved the javascript script for returning the json file from the html to a new file in the static folder

G-P-x committed:
modified the classification_output.html adding some CSS and moving the buttons in another position to have a better look

G-P-x committed:
modified the button's id from downloadButton to downloadPlotButton

G-P-x committed:
modified the button'id in the CSS and changed the background-color when mouse hover

G-P-x committed:
I adjusted the layout of the file removing comments and void lines

G-P-x committed:
deleted vscode folder, it shouldn't be in the online repository

G-P-x committed:
resolve the position of </div>

G-P-x committed:
remove the additional space in main.py

G-P-x committed:
Merge pull request #10 from G-P-x/Download_results_button_branch

Download results button branch

murali-tr committed:
image histogram

murali-tr committed:
Merge pull request #21 from G-P-x/Image_Histogram_branch

image histogram

Commit:
Almost implemented upload_image task

Implemented the upload_image task in this way:
- Added an Upload button in order to select an image from the PC, and
  handled any other case that the file is not a jpg, jpeg or png file.
- Implemented the classification_select_image in order to classify the
  image loaded;
- Implemented the "Try with uploaded image!" button in base.html in
  order to redirect the user in uploading the image;
- Implemented the functions to upload and remove the images in the
  image_uploader.py file;
- Implemented the removal of the image when the user try to upload
  another image (click on the try with uploaded image button or Upload
  button);
- Setted the path to save the image loaded in the same folder of the images

Commit:
Move the remove_uploaded_image function

I moved the "remove_uploaded_image" function into the "/classifications" get
in order to classify the uploaded image and not have any problem of
overwriting with multiple users for the image uploaded

Commit:
Merge release branch into Upload_image_button_branch

andrealabelli committed:
Merge pull request #24 from G-P-x/Upload_image_button_branch

Upload image button branch

G-P-x committed:
merge release into Download_results_button_branch

G-P-x committed:
Added the Download Graph and Download Scores to the HTML file.

The classification_output_from_upload.html file didn't have this capability
to download the results, so I added the buttons to do that there too.

miche9 committed:
Image transformation feature implemented

G-P-x committed:
Merge pull request #26 from G-P-x/Download_results_button_branch

Download results button branch

miche9 committed:
Merge branch 'release' into Image_Transformation_branch

miche9 committed:
Merge pull request #27 from G-P-x/Image_Transformation_branch

Image transformation feature implemented

G-P-x committed:
update opencv-python to the requirements

G-P-x committed:
Added changelog for this version

miche9 committed:
Image transformation function modified

The image transformation function has been modified so that transformed images are saved in a temporary folder and, therefore, do not appear in the image selection menu. Each time a transformation is requested for a new image, the transformed image is overwritten. There are no issues in case of a new request from another user because, even if the previous user is still viewing the image during a new request, the previously requested image remains visible on the web page until it is closed.

miche9 committed:
merge release into my branch

miche9 committed:
Merge pull request #28 from G-P-x/Image_Transformation_branch

Image transformation branch

Commit:
Added remove_uploaded_image when calling image_histogram and modified image_uploader.py

Commit:
Merge branch 'release' into Upload_image_button_branch

andrealabelli committed:
Merge pull request #29 from G-P-x/Upload_image_button_branch

Added remove_uploaded_image

Commit:
Update changelog

andrealabelli committed:
Merge pull request #30 from G-P-x/Upload_image_button_branch

Modified changelog

G-P-x commented:
We encountered issues with merging the various branches. The release working tree became completely messed up, and we had to restore the work by starting over from the beginning. This time, we tried to follow the instructions given in the previous review, adding a changelog with the various modifications made compared to the initial project.

asotgiu reviewed (changes requested):
I rewieved again the entire code, and now all the requested functionalities are implemented.
The documentation was improved, and a changelog was added.
Image transformation looks good now, and all files that didn't require to be versioned were removed.
The concurrency problem has been solved for the results and the plot to be downloaded, but it still remains for the uploaded and transformed images.
In these two cases, you always save an image in the same file using hardcoded paths (`uploaded_image.JPEG` and `transformed_image.JPEG`, respectively). Imagine what could happen if two users upload an image nearly simultaneously. The first uploaded image might be deleted or overwritten by the second one, and both users might receive the results related to the second image (or an error). The same could happen when applying transformations to an image.
Please fix this issue to get the PR merged. I can suggest two solutions, but you are free to implement a different one:
- store images with a unique file name for each user;
- avoid storing images at all, reusing them immediately.
If you need further help, please contact us by email.
Thank you!

miche9 committed:
Modified the image_transformation function to save transformed images with a unique id

Commit:
Implemented uuid image for the uploaded image

G-P-x commented:
We solved the concurrency problem using the uuid library, It should be ok now

asotgiu reviewed (approved):
Well done, I merge the PR
